NIFI-8126 Include Total Queued Duration in ConnectionStatus metrics

JonathanKessler opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Adds total queued duration and max queued duration for ConnectionStatus objects based on the active portion of each FlowFileQueue per NIFI-8126. This allows these metrics to be exposed to ReportingTasks.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message? Yes, NIFI-8126

- [X ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? N/A
- [X] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`? N/A
- [X] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`? N/A
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties? N/A

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered? N/A

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

markobean commented:
Individual Reporting Tasks will need to be updated to include these new statistics. Otherwise, this looks good to me.
+1

markap14 commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/SwappablePriorityQueue.java:
```diff
@@ -920,7 +921,14 @@ public SwapSummary recoverSwappedFlowFiles() {
 
+    public long getMaxActiveQueuedDuration(long fromTimestamp) {
+        // We want the oldest timestamp, which will be the min
+        return fromTimestamp - activeQueue.parallelStream().map(FlowFile::getLastQueueDate).filter(Objects::nonNull).min(Long::compareTo).orElse(fromTimestamp);
```
Unfortunately, these numbers are not accurate if any FlowFiles have been swapped out. To get the max active queued duration, we would have to also consider any data in swap files. To do that efficiently, before swapping data out, the oldest timestamp would need to be written to the swap file header. Even scanning the header, though, would be too expensive to invoke regularly, so it would be important that this class also keep a mapping of swap file to earliest timestamp, and recover that on startup when recovering swapped data.

Additionally, we cannot access the `activeQueue` outside of a Read/Write lock, so that would need to be added in here for appropriate thread safety. Would also say that we should avoid any parallel streams here. The act of creating the parallel stream is almost guaranteed to be far more expensive than the cost of just iterating over all of the elements in the queue using a for-each.

JonathanKessler replied:
Understood and agreed. I will update accordingly.

JonathanKessler replied:
@markap14 Do you have a recommendation as to how to persist & retrieve the map of swapped queueDates or how to recreate it on startup?

JonathanKessler replied:
Disregard my previous question, I misunderstood how swap files worked. I now understand that all flowfiles are swapped in from a given swap file at the same time vs one at a time. Therefore we only need to know what the min last queued date is, as well as the total of all last queued dates and total number of swapped flow files. From that we can extrapolate anything else we may need.

markap14 commented on nifi-framework-api/src/main/java/org/apache/nifi/controller/queue/FlowFileQueue.java:
```diff
@@ -93,6 +93,18 @@
+     * @param fromTimestamp The timestamp in milliseconds from which to calculate durations. This will typically be the current timestamp.
+     * @return the maximum time in milliseconds that any FlowFile within this queue has been in this queue.
+     */
+    long getMaxActiveQueuedDuration(long fromTimestamp);
```
I'm not sure what exactly it means for the max active queued duration to take into account a timestamp... it makes a lot more sense to me to accept no arguments here at all, and to instead just use the current time. Or, alternatively, to return the timestamp of the longest-queued FlowFile.

JonathanKessler replied:
It is definitely more efficient to return the timestamp and do one calculation vs calculating the difference for every flow file prior to returning a max. I will update accordingly.

markap14 commented on nifi-framework-api/src/main/java/org/apache/nifi/controller/queue/FlowFileQueue.java:
```diff
@@ -93,6 +93,18 @@
+     * @param fromTimestamp The timestamp in milliseconds from which to calculate durations. This will typically be the current timestamp.
+     * @return the sum in milliseconds of how long all FlowFiles within this queue have currently been in this queue.
+     */
+    long getTotalActiveQueuedDuration(long fromTimestamp);
```
I'm not really sure what exactly this metric is telling us. How would this metric be used? This would also become very expensive to calculate with data that is swapped out...

JonathanKessler replied:
This was an attempt to restrict the scope of this particular metric to flow files that were not swapped but in hindsight that isn't necessarily useful so will update the logic to include them as well per your suggestion. This metric is telling us the sum of the queued duration of all flow files within a given connection. My use case is to track both the average as well as the max duration to find bottlenecks/issues. I chose to report the total vs calculating and reporting an average to match the logic of other metrics contained within the ConnectionStatus class (InputBytes and OutputBytes).

markap14 reviewed (changes requested):
Thanks for the PR @JonathanKessler! I like the idea. It's definitely a good feature to add. We do have to ensure, though, that proper thread safety is observed, that it remains extremely efficient (as this method is called a lot by background threads!) and that data that is swapped out is properly considered.

JonathanKessler commented:
@markap14 I just pushed a commit that I believe addresses everything. It was a bit more involved than I anticipated so please let me know if I missed anything.

JonathanKessler commented:
@markap14, just wanted to ping you to please take a second look when you have a moment.

JonathanKessler commented:
@markap14 Pinging again, can you please take another look?

markobean commented:
@markap14 we are going to be cutting a release of an internal fork, and we'd like to include this ticket. Do you have any additional comments? Or, if it looks good, could you please commit? Thanks!

JonathanKessler commented:
> Thanks for the update @JonathanKessler . I do think that you addressed the concerns that I had in the first review. Looked over it again. There are a few places where the code should probably be cleaned up a bit (commented inline). Small changes to just make it easier to understand.
> 
> There's also the potential for a NullPointerException that must be addressed before we can merge this.
> 
> However, the stated goal of NIFI-8126 is to include a new metric about average queue duration, etc. in the stats history. This does not accomplish that. It updates the ConnectionStatus, but it doesn't give us anything in the UI to select the Average Queue Duration. The ConnectionStatusDescriptor needs to be updated to expose that information.

I will make the requested changes. The ticket actually says to expose a total or an average. I went with the total, leaving it up to consuming systems to calculate the average. This matches other metrics available within ConnectionStatus such as inputBytes, queuedBytes, etc. I will look at adding that to the ConnectionStatusDescriptor.

JonathanKessler commented:
@markap14 @markobean I just pushed a commit that I believe covers all outstanding items.

markap14 commented:
Thank you @JonathanKessler! This is a great improvement. I can also see a future improvement that would be great where we could use this information in the Status page in order to sort queues by the Average Queue Time, etc. so that it will help to determine where bottlenecks may be in the flow, etc.

The only issue I had with the update was the verbiage used in the ConnectionStatusDescriptor (these stats are not related to the last 5 minutes). So I tweaked the descriptions there, changed the type of Max Queued Duration to DURATION instead of COUNT. Also added in the Average in addition to the max, so that it's provided in the NiFi UI (this won't be shown in the Connection Status, etc. only in the UI).

Otherwise, all looks really good. This was a very non-trivial improvement, and I appreciate you sticking with us to get everything sorted. Great contribution, thanks again!

JonathanKessler commented:
Happy to contribute, thank you for the guidance!
